src/main/java/org/crac/Core.java

```diff
@@ -103,7 +103,7 @@ public String toString() {
         }
     }
 
-    static abstract class Compat {
+    static class Compat {
```

Comment:
naming: it would nice to find a more descriptive names for Compat and Core classes. 

Core is too generic, it would be valuable to pick a more specific name.
Compat name can be more description, which highlights functional role of the class.

Reply:
We'd need to stick with the `Core` that mirrors https://crac.github.io/openjdk-builds/javadoc/api/java.base/jdk/crac/Core.html. Or it should be fixed in the OpenJDK CRaC first.

Reply:
Understood. Is it hard to make changes in OpenJDK CRaC?

I don't have any strong opinions at the moment, just trying to explore different approaches to make different CRaC Core implementations discoverable and extensible.

For example - would it be more flexible to re-use Java's SPI to resolve the CRaC implementation.
https://www.baeldung.com/java-spi

Reply:
> Understood. Is it hard to make changes in OpenJDK CRaC?

It should be possible, but it's not very cheap. After a change in the JDK, org-crac 1.x should be teached to use the new API in the JDK. As then as the eventual goal, org-crac 2.x that mirrors the new API should be released.

> I don't have any strong opinions at the moment, just trying to explore different approaches to make different CRaC Core implementations discoverable and extensible.
> 
> For example - would it be more flexible to re-use Java's SPI to resolve the CRaC implementation.

This would be probably fine as internal implementation detail for org-crac, e.g. switch from reflection to the Service-based discovery. Although some Service Providers may have to use reflection to understand if they support the particular JDK.

The public interface of org-crac does not fit into Service/Provider model. "A service is a single type, usually an interface or abstract class" (https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html), and org-crac provides a bunch of classes. That, and also we don't force users to select a proper implementation.

Reply:
On SPI - I was thinking we can introduce a CracContextProvider, which would give you a Context implementation to subscribe to. Clearly it will be a breaking change to the existing crac APIs, hence we can punt it till crac 2.x.

Overall, I agree it's reasonable to preserve backward compatibility and keep Core class at the top. Current CR is an proposal to make Core more configurable, the question is - should we make it even more flexible/abstract about the supported implementations or not. Additional question - should we allow to completely disable crac or load a no-op implementation if a requested one doesn't exist.

Reply:
> On SPI - I was thinking we can introduce a CracContextProvider, which would give you a Context implementation to subscribe to. Clearly it will be a breaking change to the existing crac APIs, hence we can punt it till crac 2.x.

org-crac does not need to expose Service API, that is, expose that there are multiple implementations exist -- this will make it different with the JDK CRaC API, which we'd like everyone using. But SPI is fine as an internal implementation detail.

> Overall, I agree it's reasonable to preserve backward compatibility and keep Core class at the top. Current CR is an proposal to make Core more configurable, the question is - should we make it even more flexible/abstract about the supported implementations or not.

We've discussed an "in-memory" implementation of org-crac that does not depend on CRaC JDK. The impl can keep track of Resources and allow to e.g. call beforeCheckpoint/afterRestore. In theory, that can be made into an external project (with ServiceProvider for example), if you think it won't be useful for anyone else. But otherwise, if it is general enough, the in-memory impl may be added to org-crac. So the answer whatever works for you.

> Additional question - should we allow to completely disable crac or load a no-op implementation if a requested one doesn't exist.

The current implementation is fine if no usable Compat is found, it is no-op mode. The `register()` checks the compat and do nothing if it is missing. So no notifications won't arrive as well. The PR does not change this, so it will be no-op if the implementation specified by the system property is not found. https://github.com/CRaC/org.crac/blob/553f2e4931ac74c320e4c86eadf0d3cc4af19342/src/main/java/org/crac/Core.java#L243